### DISCUSSION TITLE
EIS for coupled-degradation

### DISCUSSION THREAD

**Discussion by Akila1993** [pybamm-team/PyBaMM#3507] [Q&A] [answered]
Hi All, 

I have tried to obtain EIS using PyBaMM-EIS for the coupled-degradation model proposed [here ](https://github.com/pybamm-team/PyBaMM/blob/develop/docs/source/examples/notebooks/models/coupled-degradation.ipynb) by @DrSOKane . The code for the said scenario is mentioned below. 

And it gives the following error: **SolverError: Cannot use ODE solver 'Scipy solver (BDF)' to solve DAE model**

Can someone guide me in this regard please?

import pbeis
import pybamm
import numpy as np
import matplotlib.pyplot as plt
import time as timer
from scipy.fft import fft

# Set up
options = {
        "thermal": "lumped",
        "surface form": "differential",
        "SEI": "solvent-diffusion limited",
        "SEI porosity change": "true",
        "lithium plating": "partially reversible",
        "lithium plating porosity change": "true",  # alias for "SEI porosity change"
        "particle mechanics": ("swelling and cracking", "swelling only"),
        "SEI on cracks": "true",
        "loss of active material": "stress-driven",
        "calculate discharge energy": "true",  # for compatibility with older PyBaMM versions
          }

model = pybamm.lithium_ion.DFN(options,name="DFN",)
param = pybamm.ParameterValues("OKane2022")

frequencies = np.logspace(-4, 2, 30)

# Time domain
I = 50 * 1e-3
number_of_periods = 20
samples_per_period = 16

def current_function(t):
    return I * pybamm.sin(2 * np.pi * pybamm.InputParameter("Frequency [Hz]") * t)

param["Current function [A]"] = current_function

start_time = timer.time()

sim = pybamm.Simulation(model, parameter_values=param,solver=pybamm.ScipySolver())

impedances_time = []
for frequency in frequencies:
    # Solve
    period = 1 / frequency
    dt = period / samples_per_period
    t_eval = np.array(range(0, 1 + samples_per_period * number_of_periods)) * dt
    sol = sim.solve(t_eval, inputs={"Frequency [Hz]": frequency})
    # Extract final two periods of the solution
    time = sol["Time [s]"].entries[-3 * samples_per_period - 1 :]
    current = sol["Current [A]"].entries[-3 * samples_per_period - 1 :]
    voltage = sol["Terminal voltage [V]"].entries[-3 * samples_per_period - 1 :]
    # FFT
    current_fft = fft(current)
    voltage_fft = fft(voltage)
    # Get index of first harmonic
    idx = np.argmax(np.abs(current_fft))
    impedance = -voltage_fft[idx] / current_fft[idx]
    impedances_time.append(impedance)

end_time = timer.time()
time_elapsed = end_time - start_time
print("Time domain method: ", time_elapsed, "s")

# Frequency domain
methods = ["direct"]
impedances_freqs = []
for method in methods:
    start_time = timer.time()
    eis_sim = pbeis.EISSimulation(model, parameter_values=parameter_values)
    impedances_freq = eis_sim.solve(frequencies, method)
    end_time = timer.time()
    time_elapsed = end_time - start_time
    print(f"Frequency domain ({method}): ", time_elapsed, "s")
    impedances_freqs.append(impedances_freq)

# Compare
_, ax = plt.subplots()
ax = pbeis.nyquist_plot(impedances_time, ax=ax, label="Time", alpha=0.7)
for i, method in enumerate(methods):
    ax = pbeis.nyquist_plot(
        impedances_freqs[i], ax=ax, label=f"Frequency ({method})", alpha=0.7
    )
ax.legend()
plt.suptitle(f"{model.name}")
plt.show()

**Comment by TomTranter** [comment]
Try removing the solver=pybamm.ScipySolver() from the Simulation. It should default to Casadi and that can solve DFN model

**Reply by rtimms** [reply]
see also https://github.com/pybamm-team/PyBaMM/discussions/2303 for a discussion on how different solvers perform on this kind of problem

**Reply by Akila1993** [reply]
@TomTranter Thank you for your suggestion, I think Casadi will not work for PyBaMM EIS according to [this discussion ](https://github.com/pybamm-team/PyBaMM/discussions/2974#discussioncomment-5965228). 

Also, I tried with the default setting it gave me the following error which confirms the above discussion. Any comments on this?

SolverError: Maximum number of decreased steps occurred at t=0.0 (final SolverError: '.../casadi/interfaces/sundials/idas_interface.cpp:591: IDACalcIC returned "IDA_NO_RECOVERY". Consult IDAS documentation.'). For a full solution try reducing dt_max (currently, dt_max=631.2500000000146) and/or reducing the size of the time steps or period of the experiment. Set `return_solution_if_failed_early=True` to return the solution object up to the point where failure occured.

**Reply by Akila1993** [reply]
@rtimms if you refer the[ previous discussion](https://github.com/pybamm-team/PyBaMM/discussions/2974#discussioncomment-5965228) we cannot used Casadi with PyBaMM-EIS. And tried with different solvers as suggested but it seems not working PyBaMM-EIS.

**Reply by Akila1993** [reply]
Just to update: I removed following options and got the EIS profile from frequency domain section without any error. 

Just to be clear  I include each options one by in [in this code](https://docs.pybamm.org/en/latest/source/examples/notebooks/models/coupled-degradation.html) and each if following options gives the same error as mentioned above. 

- "particle mechanics": ("swelling and cracking", "swelling only"),
- "SEI on cracks": "true",
-  "loss of active material": "stress-driven",

@DrSOKane do you have any idea why it gives the above error indicated in previous comment.

**Reply by DrSOKane** [reply]
Hi Akila,

I have never tried running EIS in PyBaMM so I do not know why the mechanics model is causing errors.

I can tell you that the cracking model is much harder to solve than the swelling model, so I would start with `"swelling only"` and `"stress driven"`, leaving SEI on cracks for later.

I can also tell you that the mechanics model depends on concentration gradients, which will show... interesting behaviour for an EIS experiment. Try reducing the number of mesh points in the particle domain (default is 20). If that doesn't work, try increasing it.

This will involve a lot of trial and error, so set up your models to fail fast. Try reducing the number of points in the x direction. 10 each for negative, separator and positive domains is normally more than enough for this parameter set, but as I said I've never tried EIS. You can also turn off `"calculate discharge energy"`. I only turned it on for compatibility with older versions.

Please let tme know if altering the mesh size solves the problem. If it does not, it could be the assumptions behind the mechanics model simply don't apply during EIS experiments!

Keep going!
Simon

**Reply by Akila1993** [reply]
Hi @DrSOKane ,

I have included all the changes and it gives me the promising results (Did not include SEI on cracks for now). And also as you mentioned the EIS behaviour is quite interesting with coupled degradation models. 

Thanks for your time and invaluable comments and advises.

**Reply by Akila1993** [reply]
Hello @DrSOKane ,

Is it possible use other parameter sets such as Chen2020 with for the following options. 

```python3
model = pybamm.lithium_ion.DFN(
    {   
        "thermal": "lumped", #lumped thermal submodel 
        "surface form": "differential", #Adding capacitance to the model 
        "SEI": "solvent-diffusion limited",
        "SEI porosity change": "true",
        "lithium plating": "partially reversible",
        "lithium plating porosity change": "true",  # alias for "SEI porosity change"
        "particle mechanics":"swelling only",
        "SEI on cracks": "true",
        "loss of active material": "stress-driven",
        
    }
)
```

I am getting errors with lithium plating options and subsequently we turn of plating option we cannot get "particle mechanics", "SEI on cracks": "loss of active material": too. 

Please advise. 

Thanks in advance.

**Reply by brosaplanella** [reply]
Some model combinations might be incompatible, can you please paste the full error message to check if that's the case?

**Reply by DrSOKane** [reply, marked as answer]
`OKane2022` is the only ready-made parameter set that works with all degradation options. It was designed for that purpose.

Fortunately, you can add degradation parameters to any parameter set by making a copy of the parameter file and copy-pasting the degradation parameters you need from `OKane2022`. You can then load the parameters using

```
from your_parameter_set import get_parameter_values
parameter_values = get_parameter_values()
```

**Reply by Akila1993** [reply]
@DrSOKane Thank you for your input. It works!
